connector/saml/saml.go

```diff
@@ -0,0 +1,300 @@
+}
+
+// Open validates the config and returns a connector. It does not actually
+// validate connectivity with the provider.
+func (c *Config) Open(logger logrus.FieldLogger) (connector.Connector, error) {
+	return c.openConnector(logger)
+}
+
+func (c *Config) openConnector(logger logrus.FieldLogger) (interface {
+	connector.SAMLConnector
+}, error) {
+	requiredFields := []struct {
+		name, val string
+	}{
+		{"issuer", c.Issuer},
+		{"ssoURL", c.SSOURL},
+		{"usernameAttr", c.UsernameAttr},
+		{"emailAttr", c.EmailAttr},
+		{"redirectURI", c.RedirectURI},
+	}
+	var missing []string
+	for _, f := range requiredFields {
+		if f.val == "" {
+			missing = append(missing, f.name)
+		}
+	}
+	switch len(missing) {
+	case 0:
+	case 1:
+		return nil, fmt.Errorf("missing required field %q", missing[0])
+	default:
+		return nil, fmt.Errorf("missing required fields %q", missing)
+	}
+
+	p := &provider{
+		issuer:       c.Issuer,
+		ssoURL:       c.SSOURL,
+		now:          time.Now,
+		usernameAttr: c.UsernameAttr,
+		emailAttr:    c.EmailAttr,
+		groupsAttr:   c.GroupsAttr,
+		redirectURI:  c.RedirectURI,
+		logger:       logger,
+	}
+
+	if !c.InsecureSkipSignatureValidation {
+		if (c.CA == "") == (c.CAData == nil) {
+			return nil, errors.New("must provide either 'ca' or 'caData'")
+		}
+
+		var caData []byte
+		if c.CA != "" {
+			data, err := ioutil.ReadFile(c.CA)
+			if err != nil {
+				return nil, fmt.Errorf("read ca file: %v", err)
+			}
+			caData = data
+		} else {
+			caData = c.CAData
+		}
+
+		var (
+			certs []*x509.Certificate
+			block *pem.Block
+		)
+		for {
+			block, caData = pem.Decode(caData)
+			if block == nil {
+				break
+			}
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, fmt.Errorf("parse cert: %v", err)
+			}
+			certs = append(certs, cert)
+		}
+		if len(certs) == 0 {
+			return nil, errors.New("no certificates found in ca data")
+		}
+		p.validator = dsig.NewDefaultValidationContext(certStore{certs})
+	}
+	return p, nil
+}
+
+type provider struct {
+	issuer string
+	ssoURL string
+
+	now func() time.Time
+
+	// If nil, don't do signature validation.
+	validator *dsig.ValidationContext
+
+	usernameAttr string
+	emailAttr    string
+	groupsAttr   string
+
+	redirectURI string
+
+	logger logrus.FieldLogger
+}
+
+func (p *provider) POSTData(s connector.Scopes) (action, value string, err error) {
+
+	// NOTE(ericchiang): If we can't follow up with the identity provider, can we
+	// support refresh tokens?
+	if s.OfflineAccess {
+		return "", "", fmt.Errorf("SAML does not support offline access")
+	}
+
+	r := &authnRequest{
+		ProtocolBinding: bindingPOST,
+		ID:              "_" + uuidv4(),
+		IssueInstant:    xmlTime(p.now()),
+		Destination:     p.ssoURL,
+		Issuer: &issuer{
+			Issuer: p.issuer,
+		},
+		NameIDPolicy: &nameIDPolicy{
+			AllowCreate: true,
+			Format:      "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified",
+		},
+		AuthnContextClass: &authnContextClass{
```

Comment:
I think it's an invalid element. e.g. this creates

```xml
<AuthnRequest>
    <AuthnContextClassRef>
    </AuthnContextClassRef>
</AuthnRequest>
```

but the spec wants

```xml
<AuthnRequest>
    <RequestedAuthnContext>
        <AuthnContextClassRef>
        </AuthnContextClassRef>
    </RequestedAuthnContext>
</AuthnRequest>
```

let me fix that.